app/assets/javascripts/discourse/initializers/copy-codeblocks.js

```diff
@@ -0,0 +1,121 @@
+import { withPluginApi } from "discourse/lib/plugin-api";
+import { later } from "@ember/runloop";
+import { Promise } from "rsvp";
+import { iconHTML } from "discourse-common/lib/icon-library";
+
+// http://github.com/feross/clipboard-copy
+function clipboardCopy(text) {
+  // Use the Async Clipboard API when available. Requires a secure browsing
+  // context (i.e. HTTPS)
+  if (navigator.clipboard) {
+    return navigator.clipboard.writeText(text).catch(function(err) {
+      throw err !== undefined
+        ? err
+        : new DOMException("The request is not allowed", "NotAllowedError");
+    });
+  }
+
+  // ...Otherwise, use document.execCommand() fallback
+
+  // Put the text to copy into a <span>
+  var span = document.createElement("span");
+  span.textContent = text;
+
+  // Preserve consecutive spaces and newlines
+  span.style.whiteSpace = "pre";
+
+  // Add the <span> to the page
+  document.body.appendChild(span);
+
+  // Make a selection object representing the range of text selected by the user
+  var selection = window.getSelection();
+  var range = window.document.createRange();
+  selection.removeAllRanges();
+  range.selectNode(span);
+  selection.addRange(range);
+
+  // Copy text to the clipboard
+  var success = false;
+  try {
+    success = window.document.execCommand("copy");
+  } catch (err) {
+    // eslint-disable-next-line no-console
+    console.log("error", err);
+  }
+
+  // Cleanup
+  selection.removeAllRanges();
+  window.document.body.removeChild(span);
+
+  return success
+    ? Promise.resolve()
+    : Promise.reject(
+        new DOMException("The request is not allowed", "NotAllowedError")
+      );
+}
+
+let _clickHandlerElement = null;
+
+export default {
+  name: "copy-codeblocks",
+
+  initialize(container) {
+    withPluginApi("0.8.7", api => {
+      function _cleanUp() {
+        if (_clickHandlerElement) {
+          _clickHandlerElement.removeEventListener("click", _handleClick);
+          _clickHandlerElement = null;
+        }
+      }
+
+      function _handleClick(event) {
+        if (!event.target.classList.contains("copy-cmd")) {
+          return;
+        }
+
+        const button = event.target;
+        const code = button.nextSibling;
+
+        if (code) {
+          let string = code.innerText;
+
+          if (string) {
+            string = string.replace(/^\s+|\s+$/g, "");
+            clipboardCopy(string);
+          }
+
+          button.classList.add("copied");
+
+          later(() => button.classList.remove("copied"), 3000);
+        }
+      }
+
+      function _attachCommands($elem) {
+        const siteSettings = container.lookup("site-settings:main");
+        if (siteSettings.show_copy_button_on_codeblocks) {
```

Comment:
it's usually better to do:

```
if (!siteSettings.show_copy_button_on_codeblocks) {
  return;
}
```

It avoids adding one un-necessary indentation level